lib/diskoverflow/map.go

```diff
@@ -0,0 +1,315 @@
+	if !ok {
+		return false
+	}
+	copyValue(v, nv)
+	return true
+}
+
+func (o *memoryMap) Items() int {
+	return len(o.values)
+}
+
+func (o *memoryMap) Pop(k string, v Value) bool {
+	ok := o.Get(k, v)
+	if !ok {
+		return false
+	}
+	delete(o.values, k)
+	o.bytes -= v.ProtoSize()
+	return true
+}
+
+func (o *memoryMap) Delete(k string) {
+	v, ok := o.values[k]
+	if !ok {
+		return
+	}
+	delete(o.values, k)
+	o.bytes -= v.ProtoSize()
+}
+
+type iteratorValue struct {
+	k string
+	v Value
+}
+
+type memMapIterator struct {
+	values  map[string]Value
+	ch      chan iteratorValue
+	stop    chan struct{}
+	current iteratorValue
+	parent  iteratorParent
+}
+
+func (o *memoryMap) newIterator(p iteratorParent) MapIterator {
+	i := &memMapIterator{
+		ch:     make(chan iteratorValue),
+		stop:   make(chan struct{}),
+		parent: p,
+		values: o.values,
+	}
+	go i.iterate()
+	return i
+}
+
+func (i *memMapIterator) iterate() {
+	for k, v := range i.values {
+		select {
+		case <-i.stop:
+			break
+		case i.ch <- iteratorValue{k, v}:
+		}
+	}
+	close(i.ch)
+}
+
+func (i *memMapIterator) Key() string {
+	return i.current.k
+}
+
+func (i *memMapIterator) Value(v Value) {
+	copyValue(v, i.current.v)
+}
+
+func (i *memMapIterator) Next() bool {
+	var ok bool
+	i.current, ok = <-i.ch
+	return ok
+}
+
+func (i *memMapIterator) Release() {
+	close(i.stop)
+	i.parent.released()
+}
+
+type diskMap struct {
+	db    *leveldb.DB
+	bytes int
+	dir   string
+	len   int
+}
+
+func newDiskMap(location string) *diskMap {
+	// Use a temporary database directory.
+	tmp, err := ioutil.TempDir(location, "overflow-")
+	if err != nil {
+		panic("creating temporary directory: " + err.Error())
+	}
+	db, err := leveldb.OpenFile(tmp, &opt.Options{
+		OpenFilesCacheCapacity: 10,
+		WriteBuffer:            512 << 10,
+	})
+	if err != nil {
+		panic("creating temporary database: " + err.Error())
+	}
+	return &diskMap{
+		db:  db,
+		dir: tmp,
+	}
+}
+
+func (o *diskMap) set(k string, v Value) {
+	o.addBytes([]byte(k), v)
+	o.bytes += v.ProtoSize()
+}
+
+func (o *diskMap) addBytes(k []byte, v Value) {
+	d, err := v.Marshal()
+	errPanic(err)
+	if err := o.db.Put(k, d, nil); err != nil {
+		panic("writing to temporary database: " + err.Error())
+	}
+	o.len++
+}
+
+func (o *diskMap) Close() {
+	o.db.Close()
+	os.RemoveAll(o.dir)
+}
+
+func (o *diskMap) Bytes() int {
+	return o.bytes
+}
+
+func (o *diskMap) Get(k string, v Value) bool {
+	d, err := o.db.Get([]byte(k), nil)
+	if err != nil {
+		return false
+	}
+	errPanic(v.Unmarshal(d))
+	return true
+}
+
+func (o *diskMap) Items() int {
+	return o.len
+}
+
+func (o *diskMap) Pop(k string, v Value) bool {
+	ok := o.Get(k, v)
+	if ok {
+		errPanic(o.db.Delete([]byte(k), nil))
+		o.len--
+	}
+	return ok
+}
+
+func (o *diskMap) Delete(k string) {
+	errPanic(o.db.Delete([]byte(k), nil))
+	o.len--
+}
+
+func (o *diskMap) newIterator(p iteratorParent) MapIterator {
+	return &diskIterator{
+		it:     o.db.NewIterator(nil, nil),
```

Comment:
It's interesting that iterator order will be different between the memory one and disk one.

Reply:
It's a map, there's no defined order so doesn't matter (?)

Reply:
Yeah, but it seems the sorted thing and this thing could be the same.